### ISSUE TITLE
Malformed Asciimath equations 

### ISSUE THREAD

**Issue by opoudjis** [metanorma/stepmod-utils#191]
In the ongoing problems we have had getting ISO-10303-42 to generate (under https://github.com/metanorma/annotated-express/issues/137 ), we have a new issue with the recent introduction of boldface and italics formatting into Asciimath.

The problem is that the introduced formatting is disrupting the well-formedness of the generated Asciimath. Two equations are currently preventing compilation, but these kinds of issue will occur elsewhere:

First,

Express:

ID: geometry_schema.b_spline_volume

````
 <eqn> <b>P</b><i><sub>ijk</sub></i>   = <b>control_values</b>     </eqn>
````

Generated Asciimath:

````
bb(P)ii(_{ijk}) = bb("control_values")
````

should be:

````
bb(P)_ii({ijk}) = bb("control_values")
````

Second:

ID: topology_schema.face

Express:

````
eqn>
      (<b><i>V</i></b> - <b><i>E</i></b>) - (<b><i>L</i></b> - &#931;<sup>L</sup><sub><i>i=1</i></sub> (G<i><sub>i</sub><sup>l</sup></i>)) = 0 &#8195; &#8195; &#8195; &#8195;<A name="eqn5"> (5) </A> </eqn>
````

Generated Asciimath:

````
bb(ii(V)) - bb(ii(E))) - (bb(ii(L)) - Σ^{L}_{ii(i=1)} (Gii(_{i}^{l}))) = 0
````

should be:

````
(bb(ii(V)) - bb(ii(E))) - (bb(ii(L)) - Σ^{L}_{ii(i=1)} (G_ii({i}^{l}))) = 0

````

**Comment by opoudjis** [comment]
@suleman-uzair There is nothing in my opinion for you to do here, but I'm including you just in case you need to intervene in the equations somehow.

**Comment by opoudjis** [comment]
There is a common issue here @HassanAkbar , which is that

````
<i><sub>ijk</sub></i>
````

needs to be transformed by *something* to

````
<sub><i>ijk</i></sub>
````

before processing, and that needs to be generalised. 

sub is a relational operator, translated into Asciimath x_y. Any italics or boldface needs to be nested inside it. And the same in fact goes for any such operator.

When this kind of overlap of hierarchies has come up in Metanorma, what I have ended up doing is putting the new markup at the deepest level of the tree, to ensure it doesn't disrupt the existing hierarchies. So I do this:

````
x.traverse do |x|
  next unless x.text?
  x.replace("<i>#{x.text}</i>")
end
````

That means that, the minute I see a font, in this case, I don't wrap the entire subexpression in the font, because I simply can't trust that it won't destroy the integrity of the hierarchy. Instead, I wrap each and every *text node*, each leaf within the tree with the font command.

**Comment by HassanAkbar** [comment]
@opoudjis I have a question related to this, for example if we have the following equation 

```xml
<eqn id="eqnGM1">
  <i>
    &#967;
    <sub>ms</sub>
    =
    <b>V - E + 2F - L</b>
    <sub>l</sub>
    <b> - 2(S - G</b>
    <sup>s</sup>)
  </i>
  = 0 &#8195; (1) &#8195;
</eqn>
```

If we move `<I>` tag to child nodes it will become something like

```xml
<eqn id="eqnGM1">
  <i>&#967;</i>
  <sub><i>ms</i></sub>
  <i>=</i>
  <b><i>V - E + 2F - L</i></b>
  <sub><i>l</i></sub>
  <b><i> - 2(S - G</i></b>
  <sup><i>s</i></sup>
  <i>)</i>
  = 0 &#8195; (1) &#8195;
</eqn>
```

If we convert this to asciimath the last closing bracket inside the `<i>` tag will become `ii())` for fixing this issue we can add double quotes inside the `ii` like `ii(")")`.
My question is will it cause any problem to add double quotes inside the `ii` or `bb` modifier everywhere? or should I only add this if we have brackets inside the modifiers?

**Comment by opoudjis** [comment]
I see the problem. Ugh. No, ")" won't work, because it will mean that its contents will become text, and opaque to Asciimath.

In fact, that expression is unsalvageable in its current form, because the closing bracket will be detached from the opening bracket, and Asciimath won't be able to put them back together.

You're going to have to do even more preprocessing I'm afraid, and we're going to have to take a different approach:

* Bold and italics that apply to an entire equation remain at that high level, so they don't introduce more problems
* Bold and italics inside an equation need to be passed inside of any existing XML tags like sub
* But in addition, bold italics that wrap text other than alphanumerics need to be ignored, and bold italics that wrap text including both alphanumerics and non-alphanumerics need to be wrapped around only alphanumeric content.

So:

````
<b> - 2(S - G</b>
    <sup>s</sup>)
```` 

needs to be converted into:

````
- 2(<b>S<b> - <b>G</b><sup>s</sup>)
````

Otherwise the Ascimath simply won't work: you'll get the unparseable bb(- 2(S -G)_s. If you do the manipulation above, you'll get -2(bb(S) - bb(G)_s), which is parseable.

I know this is going to blow out the timelines, cc @TRThurman, but that's what happens when the source material is done sloppily.

If @suleman-uzair can help with the cleanup, he should, but this cleanup does need to happen before the Asciimath parser is invoked.

**Comment by ronaldtse** [comment]
The math conversion issues are frankly a waste of resources. The important thing is we are alerted to, and move away from broken math. It’s good enough just to know where things are broken and we fix later. We CANNOT fix broken math automatically. 

For now, drop all the italics and bolding from the original equations.

**Comment by ronaldtse** [comment]
We are already paying through our nose to migrate the SMRL. *Rubbish in rubbish out*: broken source content is not our problem.

**Comment by TRThurman** [comment]
Thanks @ronaldtse!
It is not your job to fix the broken math source.
I will bring this up to the part 42 team.

**Comment by TRThurman** [comment]
@opoudjis @ronaldtse , I extracted some content from Nick's post above to add to a new bug in wg12-stepdev: https://sd.iso.org/jira-pilot/browse/TCSC410303-2050. It will be fixed for the next publication.

**Comment by HassanAkbar** [comment]
> For now, drop all the italics and bolding from the original equations.

@ronaldtse Sure, I'll open a PR to revert all the changes related to bold and italics.

**Comment by opoudjis** [comment]
@HassanAkbar , I am concerned that if we remove the bold and italics in equations during translation, the maintainers of the standard will not be able to put them back in if they want to.

It is also the case that some current equations will work when translated to AsciiMath and MathML, and some will not; I think we would be adding value to ISO-10303 if we track them.

It also needs to be said that Metanorma follows the MathML default of italics for variables, and additional settings will be needed to default to plain instead of italics for variables, as seems to be implied in the current formulas. This is feasible as a setting, but it will not happen out of the box.

I propose the following course of action, @HassanAkbar ; @ronaldtse has already approved it (at least in principle):

1. Log all equations encountered containing boldface and italics, during translation of Express to Asciidoc, in a file that maintainers can access, with the following content:

* The raw Express formula, with boldface and italics tags
* The document it is in
* The nearest Express anchor, so we can find the equation in the converted text
* The Asciimath conversion of the Express formula, *including* bold and italics
* (Possibly) The Asciimath conversion of the Express formula, *without* bold and italics, which is what is actually going into the Asciidoc
* The status of whether that Asciimath formula including bold and italics is valid according to Plurimath: if it is not, that indicates that it needs to be fixed.


To do that validation, you will do what I am already doing in https://github.com/metanorma/metanorma-standoc/blob/f25f7d10f2c42c0772207a9104e02aa90a17648c/lib/metanorma/standoc/validate.rb#L34 ; I've changed the code to match what you will be doing.

````
     def math_validate(expr)
          Plurimath::Math.parse(expr, "asciimath").to_mathml
        rescue StandardError => e
          math_validate_error(math, m, e)
      end
````

So we take the asciimath, parse it into Plurimath's internal representation, and then generate MathML (as Metanorma eventually will). If something blows up, we want to know about it.


2. Quite separately from the bold and italics issue, I would like you to validate ALL Asciimath formulas you are generating that way. If any of them result in an error when they are converted to MathML, I want you NOT to include them as is in the generated Asciidoc, because they will result in the crashes we are currently seeing. Instead, I want you to escape them as sourcecode, and add them as well to a logfile, so that maintainers know they need to fix them:

````
[source]
====
INVALID FORMULA:

{ Express formula }

{ Converted Asciimath }
====
````

**Comment by TRThurman** [comment]
This annotation will be a great help to the maintainers!
Much appreciated.
I just had a look to the annotated EXPRESS source for an equation in part 104. 
The tag that it is a defined language (in that case, latexmath) is also a win for the maintainer as it provides a hook for a language server to create/check syntax as it is being written in the future.
This is a huge improvement over the current situation where the XSLT defines the syntax in an ad-hoc manner.
Note: As a reminder to myself, there is a provision in ISO Directives that variables are to be italicized (plus other related provisions in that domain.) There are no overriding provisions in the SC4 supplementary directives. 
I look forward to raising these issues as technical issues against the documents.

**Comment by HassanAkbar** [comment]
> Log all equations encountered containing boldface and italics, during translation of Express to Asciidoc, in a file that maintainers can access, with the following content:

@opoudjis The asciidoc is generated by `stepmod2mn` so I'm not sure how to generate a log for that process. Still, I can generate the log file for the information available when converting XML to annotated_express files using `stepmod-utils`.

I have a question related to this point

> The Asciimath conversion of the Express formula, including bold and italics

The current implementation for adding bold and italics was reverted because it was distorting the formula. Should I add the bold and italics as they appear in the XML or should I add the implementation that you mentioned which skips the non-alphanumeric characters? If we want to go to the second route then it will take some time because it has not been implemented yet.

> Quite separately from the bold and italics issue, I would like you to validate ALL Asciimath formulas you are generating that way. If any of them result in an error when they are converted to MathML, I want you NOT to include them as is in the generated Asciidoc, because they will result in the crashes we are currently seeing. Instead, I want you to escape them as source code, and add them as well to a logfile, so that maintainers know they need to fix them:

I believe this is done in `stepmod2mn` so I can add these to the log file after validating but I have no idea about how Asciidoc files are generated.

**Comment by opoudjis** [comment]
OK, as you know, @HassanAkbar, I don't know how the conversion works, but I am referring here to the translation from Express markup to Asciidoc markup, which is being done by stepmod-utils, specifically the conversion of Express equations into Asciimath. Inasmuch as you are currently doing that in  stepmod-utils, that conversion needs to be logged. Obviously if stepmod2mn ends up rearranging the files you generate, you can't anticipate that in your own logging, and you can only log against the files you generate. But as it turns out, as far as I can tell, the *.exp files you generate are left alone by stepmod2mn, and are the source read in by Lutaml in Metanorma execution; and they are the main source of equations.

The bold and italics will eventually be added back in, because they are part of what the ISO-10303 team want the equations to look like. But they will be added back in manually. That's why I want you to log BOTH the equation with the bold and italics stripped, AND with the bold and italics left in. Leave the implementation I suggested to the side for the moment; I think it is the right thing to do, but that isn't urgent, and it can always be generated later from the Express originals, which you are also capturing in the log.

As far as I can tell, you and not @Intelligent2013 are doing the conversion of equations to Asciimath, and putting that in markup inside of annotations, that can only be characterised as Asciidoc markup. You are generating markup like this in iso-10303-stepmod, are you not?

https://github.com/metanorma/iso-10303-stepmod/blob/14e0bd121a8bb7ef47f0c8811e5a70dce660ebd1/data/resources/topology_schema/topology_schema_annotated.exp#L1600C1-L1616C1

````
(*"topology_schema.loop"
A **loop** is a type of <<express:topology_schema.topological_representation_item,topological_representation_item>>, constructed from a single vertex, or by stringing together connected (oriented) edges, or linear segments beginning and ending at the same vertex. A loop has dimensionality 0 or 1. The domain of a 0-dimensional loop is a single point. The domain of a 1-dimensional loop is a connected, oriented curve, but need not be a manifold. As the loop is a cycle, the location of its beginning/ending point is arbitrary. The domain of the loop includes its bounds, and 0 ≤ Ξ \< ∞

A loop is represented by a single vertex, or by an ordered collection of <<express:topology_schema.oriented_edge,oriented_edge>>s, or by an ordered collection of points.

A loop is a graph, so M and the graph genus G^_l_^ may be determined by the graph traversal algorithm. Since M = 1, the Euler equation reduces in this case to

[[eqn4]]

[stem]
++++
(V - E_{l}) - (1 - G^{l}) = 0
++++

where **_V_** and **_E_** are the number of unique vertices and oriented edges in the loop and G^_l_^ is the genus of the loop.
*)
````

That markup is not an Asciidoctor document, since it is embedded inside of Express, but it is decidedly Asciidoctor markup. What I am requesting is that you not just insert the `[stem]` expression you're generating in stepmod-utils without checking it. If the Asciimath you generate is not valid Asciimath, as far as Plurimath is concerned, I want you NOT
to insert an expression like

````
[stem]
++++
(V - E_{l}) - (1 - G^{l}) = 0
++++
````

but an expression like

````
[source]
====
INVALID FORMULA:

V - E<sub>l</sub> - (1 - G<sup>l</sup>) = 0

(V - E_{l}) - (1 - G^{l}) = 0
====
````

If Alex and not you is generating that `[stem]` expression in the exp file, then Alex should also be doing that validation, but I think it's stempod-utils that is doing that.

**Comment by HassanAkbar** [comment]
> that can only be characterised as Asciidoc markup. You are generating markup like this in iso-10303-stepmod, are you not?

@opoudjis Got it now, I misunderstood this and thought that you were talking about asciidoc documents.

**Comment by HassanAkbar** [comment]
@opoudjis I've added the logs in this PR -> https://github.com/metanorma/stepmod-utils/issues/229

Is there anything else that needs to be done in this ticket or should I close this now?

**Comment by opoudjis** [comment]
Close it, I think. The conversions will clearly need more work, but absent a strategy, we don't want an umbrella ticket like this.
